pkgs/binary-merkle-tree-halo2/src/binary_merkle_tree.rs

```diff
@@ -0,0 +1,264 @@
+                .collect::<Vec<u8>>(),
+            path_indices: self
+                .path_indices
+                .iter()
+                .flat_map(|path_index| F::from(*path_index as u64).to_bytes_le())
+                .collect::<Vec<u8>>(),
+            root: self.root.to_bytes_le(),
+        })
+    }
+
+    // TODO: impl from_bytes_le()
+}
+
+// TODO: maybe adding PoseidonConstants in the PSE version
+impl<'a, F, const T: usize, const RATE: usize, const ARITY: usize>
+    BinaryMerkleTree<'a, F, T, RATE, ARITY>
+where
+    F: BigPrimeField,
+{
+    pub fn new(poseidon: &'a mut Poseidon<F, T, RATE>) -> Self {
+        Self {
+            root: F::ZERO,
+            leaves: Vec::new(),
+            poseidon,
+            is_tree_ready: false,
+            layers: Vec::new(),
+            depth: 0,
+        }
+    }
+
+    /// Add the leaf to the bottom-most layer
+    pub fn insert(&mut self, leaf: F) {
+        self.leaves.push(leaf);
+    }
+
+    fn hash(poseidon: &mut Poseidon<F, T, RATE>, nodes: &[F]) -> F {
+        poseidon.update(nodes);
+
+        poseidon.squeeze_and_reset()
+    }
+
+    pub fn finish(&mut self) {
+        let padding_len = self.leaves.len().next_power_of_two();
+        self.leaves.resize(padding_len, F::ZERO);
+
+        let depth = (padding_len as f64).log2() as usize;
+
+        self.depth = depth;
+
+        self.is_tree_ready = true;
+
+        // Calculate the root
+        let root = self.gen_root();
+        self.root = root;
+    }
+
+    pub fn gen_root(&mut self) -> F {
+        if !self.is_tree_ready {
+            panic!("MerkleTree: Tree is not ready.");
+        }
+
+        self.layers.push(self.leaves.clone());
+        let mut current_layer = self.layers[0].clone();
+
+        (0..self.depth).for_each(|_| {
+            let layer_above = current_layer
+                .par_chunks(ARITY)
+                .map(|nodes| {
+                    let mut poseidon = self.poseidon.clone();
+                    Self::hash(&mut poseidon, nodes)
+                })
+                .collect::<Vec<F>>();
+
+            self.layers.push(layer_above.clone());
+            current_layer = layer_above
+        });
+
+        assert_eq!(current_layer.len(), 1);
+
+        current_layer[0]
+    }
+
+    pub fn gen_proof(&self, leaf: F, address: String) -> Result<MerkleProof<F>> {
+        if !self.is_tree_ready {
+            panic!("MerkleTree: Tree is not ready.");
+        }
+
+        let mut siblings = vec![];
+        let mut path_indices = vec![];
+
+        let mut current_layer = &self.layers[0];
+
+        let mut leaf_index = self
+            .leaves
+            .iter()
+            .position(|&x| x == leaf)
+            .with_context(|| {
+                format!(
+                    "Merkle proof: Leaf not found. Address {} is not part of the addresses set.",
+                    address
+                )
+            })?;
+
+        for i in 0..self.depth {
```

Comment:
shouldn't this be (not sure)
```suggestion
        for i in 0..self.depth.unwrap() -1 {
```

do we need to include the root layer in the iteration? The root layer has no siblings

Reply:
I think that is wrong because in the gen_root we loop through the whole depth to construct layers. So in gen proof we will have to loop through the whole layers.